[libc] [search] implement hcreate(_r)/hsearch(_r)/hdestroy(_r)

SchrodingerZhu opened the pull request:
This patch implements `hcreate(_r)/hsearch(_r)/hdestroy(_r)` as specified in https://man7.org/linux/man-pages/man3/hsearch.3.html.

Notice that `neon/asimd` extension is not yet added in this patch. 

- The implementation is largely simplified from rust's [`hashbrown`](https://github.com/rust-lang/hashbrown/blob/master/src/raw/mod.rs) as we only consider fix-sized insertion-only hashtables. Technical details are provided in code comments.

- This patch also contains a portable string hash function, which is derived from [`aHash`](https://github.com/tkaitchuck/aHash)'s fallback routine. Not using any SIMD acceleration, it has a good enough quality (passing all SMHasher tests) and is not too bad in speed.

- Some general functionalities are added, such as `memory_size`, `offset_to`(alignment), `next_power_of_two`, `is_power_of_two`. `ctz/clz` are extended to support shorter integers.

SchrodingerZhu committed:
[libc] [search] implement hcreate(_r)/hsearch(_r)/hdestroy(_r)

This patch implements hcreate(_r)/hsearch(_r)/hdestroy(_r) as specified in https://man7.org/linux/man-pages/man3/hsearch.3.html.

* Notice that neon/asimd extension is not yet added in this patch.

* The implementation is largely simplified from rust's hashbrown as we only consider fix-sized insertion-only hashtables. Technical details are provided in code comments.

* This patch also contains a portable string hash function, which is derived from aHash's fallback routine. Not using any SIMD acceleration, it has a good enough quality (passing all SMHasher tests at commit 7db446a0b8d2e29cd648fb5bf4224db9aed30905) and is not too bad in speed. SMHasher patch is available at https://gist.github.com/SchrodingerZhu/778a7952a35a1fc3692650b2d9fd6686.

* Some general functionalities are added, such as memory_size, offset_to(alignment), next_power_of_two, is_power_of_two, rotate_left. ctz/clz are extended to support shorter integers.

SchrodingerZhu commented:
FYI, such APIs are used in https://github.com/SchedMD/slurm/blob/e1cafb52f5041fe14a6c1af6e06243781ae994bc/src/slurmrestd/plugins/openapi/v0.0.38/jobs.c

SchrodingerZhu commented on libc/src/__support/memory_size.h:
```diff
@@ -0,0 +1,72 @@
+#else
+  T max = cpp::numeric_limits<T>::max();
+  T min = cpp::numeric_limits<T>::min();
+  bool overflow = (b > 0 && (a > max / b || a < min / b)) ||
```
This does not generate the same code as the built-in. Is there a better way to do a signed overflow check when built-in is not available?

SchrodingerZhu commented on libc/test/src/__support/HashTable/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+  UNIT_TEST_ONLY
+)
+cpu_supports(CPU_SUPPORTS_SSE2 "SSE2")
+if (CPU_SUPPORTS_SSE2)
```
I understand that there is a similar function defined in `string`. However, it is not exposed and it dispatches targets into a complicated feature matrix.

lntue replied:
You can add automated testing for SSE2 as follow:
- define `SSE2_OPT` similar to `ROUND_OPT` flag in `libc/cmake/modules/LLVMLibCFlagRules.cmake`
- expand `${SSE2_OPT_FLAG}` similar to `${ROUND_OPT_FLAG}` in `libc/cmake/modules/LLVMLibCObjectRules.cmake`, function `_get_common_compile_options`
- add `FLAGS SSE2_OPT` inside `add_header_library(bitmask ...)` in `libc/src/__support/HashTable/CMakeLists.txt` above.

Then all tests and targets depending on `libc.src.__support.HashTable.bitmask` will be extended to both with and without SSE2.

SchrodingerZhu commented on libc/utils/smhasher/smhasher.patch:
```diff
@@ -0,0 +1,205 @@
+diff --git a/Hashes.h b/Hashes.h
```
I also include a patch to `SMHasher` here to make it easier to verify the hash quality. Notice that there is ongoing work on [`SMHasher3`](https://gitlab.com/fwojcik/smhasher3/-/releases), but the author has not yet marked it as ready to use.

nickdesaulniers replied:
I think it would be better to include a link to this in the commit description (or the PR description retained when merging) than as a .patch file in the tree.  Then we don't need this or libc/utils/smhasher/smhasher.txt additions.

SchrodingerZhu replied:
Updated in latest squashed commit.

SchrodingerZhu commented on libc/src/__support/HashTable/table.h:
```diff
@@ -0,0 +1,229 @@
+
+  size_t offset_from_entries() const {
+    size_t entries_size = num_of_entries() * sizeof(ENTRY);
+    return entries_size + offset_to(entries_size, alignof(HashTable));
```
I catch this issue after the submission. It should be `table_alignment` rather than `alignof(HashTable)`.

SchrodingerZhu commented on libc/src/__support/memory_size.h:
```diff
@@ -0,0 +1,72 @@
+#include "src/__support/macros/optimization.h"
+
+namespace LIBC_NAMESPACE {
+namespace internal {
```
I am actually a little bit unhappy about the `namespace` here when I writing this patch. Maybe I should put hashtable implementation in a separated one?

SchrodingerZhu commented on libc/src/__support/hash.h:
```diff
@@ -0,0 +1,161 @@
+// Read as little endian.
+// Shift-and-or implementation does not give a satisfactory code on aarch64.
+// Therefore, we use a union to read the value.
+template <typename T> LIBC_INLINE static T read_little_endian(const void *ptr) {
```
I tested two ways of writing shift-and-or:
```c++
T res;
for(int i = 0; i < sizeof(T); ++i) {
  res = res | (bytes[i] << (i * 8));
}
```
and
```c++
T res;
for(int i = 1; i <= sizeof(T); ++i) {
  res = (res << 8) | bytes[sizeof(T) - i];
}
```
Neither of them gives satisfactory codegen with GCC (clang is fine):

The first one is very bad on both x64 and aarch64 even when being compiled along.
The second one is bad on aarch64 when used in some other code.

SchrodingerZhu commented on libc/src/__support/HashTable/randomness.h:
```diff
@@ -0,0 +1,59 @@
+namespace LIBC_NAMESPACE {
+namespace internal {
+namespace randomness {
+LIBC_INLINE_VAR thread_local static HashState state = {
```
About `LIBC_INLINE(_VAR)`: it looks weird when I write something like `static LIBC_INLINE type function()`. However, I do remember that sometimes people prefer to put storage class in the first place. What is the preferred order?

nickdesaulniers commented:
Thanks for the patch! This one will take a bit to get reviewed but we will get to it.

nickdesaulniers commented on libc/src/search/hsearch.cpp:
```diff
@@ -0,0 +1,35 @@
+#include "src/search/hsearch/global.h"
+
+namespace LIBC_NAMESPACE {
+LLVM_LIBC_FUNCTION(ENTRY *, hsearch, (ENTRY item, ACTION action)) {
+  ENTRY *result;
+
+  switch (action) {
+  case FIND:
+    result = internal::global_hash_table->find(item.key);
+    if (result == nullptr) {
+      libc_errno = ESRCH;
+    }
+    break;
+  case ENTER:
+    result = internal::global_hash_table->insert(item);
```
what happens if someone calls `hsearch` before `hcreate` has been called, or calls `hsearch` after `hcreate` then `hdestroy` have been called?

SchrodingerZhu replied:
`POSIX.1` leaves it as unspecified. Per `MUSL`'s implementation, such cases (including memory leak caused by multiple hcreate) are not checked.

SchrodingerZhu replied:
@nickdesaulniers what do you think of adding sth like `LIBC_ASSERT(internal::global_hash_table != nullptr);`?

nickdesaulniers replied:
Yeah, that's better than nothing IMO. Please add.

nickdesaulniers replied:
FWIW, I was studying bionic last night and found their implementations of these come from FreeBSD.

https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/upstream-freebsd/lib/libc/stdlib/hcreate.c

Looks like they use a `bool` to track whether the global hash table hash been initialized.  They also do the right thing if search has been called before the table was ever explicitly created.  I think we should, too.  Do you mind following up with such a change?

SchrodingerZhu replied:
@nickdesaulniers sure. Looking into the code, they are actually doing resizing and merely regard the maximum number as a hint. In that way, if a global hashtable is not initialized, they will just create a minimal table and resize it during insertion. 

It requires some modifications but will not be too painful to add such features. I will be forming a patch by next Monday.

SchrodingerZhu commented on libc/src/__support/hash.h:
```diff
@@ -0,0 +1,162 @@
+};
+
+LIBC_INLINE uint64_t rotate_left(uint64_t x, uint64_t y) {
+  return (x << y) | (x >> (64 - y));
```
Cited.
Updated to a more general version in `bit.h`

SchrodingerZhu commented on libc/src/__support/hash.h:
```diff
@@ -0,0 +1,162 @@
+  for (size_t i = 0; i < sizeof(T); ++i) {
+    data.buffer[i] = bytes[i];
+  }
+#endif
```
Not really. I only checked the codegen with related flags.

lntue commented on libc/src/__support/HashTable/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+    libc.src.__support.hash
+    libc.src.__support.common
+    ${randomness_extra_depends}
+  FLAGS
```
replace `FLAGS` with `COMPILE_OPTIONS` and move it above `DEPENDS`.

lntue commented on libc/src/__support/HashTable/bitmask.h:
```diff
@@ -0,0 +1,91 @@
+
+  // Count trailing zeros with respect to stride. (Assume the bitmask is none
+  // zero.)
+  size_t lowest_set_bit_nonzero() const {
```
I think you can mark these 2 methods `constexpr`.

lntue commented on libc/src/__support/HashTable/generic/bitmask_impl.inc:
```diff
@@ -0,0 +1,102 @@
+//    byte | (byte << 8)    = 0x000000000000ffff
+//    byte | (byte << 16)   = 0x00000000ffffffff
+//    byte | (byte << 32)   = 0xffffffffffffffff
+constexpr static inline uintptr_t repeat_byte(uintptr_t byte) {
```
Use `LIBC_INLINE constexpr uintptr_t` signature.  The macro `LIBC_INLINE` is defined in `src/__support/macros/attributes.h`

lntue commented on libc/src/__support/HashTable/sse2/bitmask_impl.inc:
```diff
@@ -0,0 +1,40 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include <emmintrin.h>
```
Including `immintrin.h` instead should be more portable, even though you only need SSE2.

lntue commented on libc/src/__support/hash.h:
```diff
@@ -0,0 +1,161 @@
+LIBC_INLINE static uint64_t folded_multiply(uint64_t x, uint64_t y) {
+  UInt128 mask = static_cast<UInt128>(0xffffffffffffffff);
+  UInt128 p = static_cast<UInt128>(x) * static_cast<UInt128>(y);
+  uint64_t low = static_cast<uint64_t>(p & mask);
```
I don't think you need the lower 64-bit mask here.  It should be well-defined in the same way as `static_cast<uint32_t>` from a `uint64_t`.

SchrodingerZhu commented:
@lntue I addressed your code reviews in latest commit. 

There is one complication: `SSE2` should be enabled on default, so I add `DISABLE_SSE2_OPT` as a negative feature flag instead.

SchrodingerZhu commented:
Should I resolve CR if it is handled? I remember I see from the discourse that some ppl are not happy about folded CRs.

lntue commented on libc/test/src/__support/HashTable/CMakeLists.txt:
```diff
@@ -30,7 +29,7 @@ if (CPU_SUPPORTS_SSE2)
       group_test.cpp
```
With the `FLAGS` option setup, you only need to create 1 test, without checking for `CPU_SUPPORT_SSE2`:
```
add_libc_test(
    group_test
    UNIT_TEST_ONLY
    SUITE
      libc-support-tests
    SRCS
      group_test.cpp
    DEPENDS
      libc.src.__support.HashTable.bitmask
      libc.src.stdlib.rand
      libc.src.string.memcmp
    DEPENDS
      libc.src.__support.HashTable.bitmask
      libc.src.stdlib.rand
      libc.src.string.memcmp
)
```
Then you after `cmake` command, you can check if the following targets are available for `ninja`:
```
  libc.test.src.__support.HashTable.group_test.__unit__
  libc.test.src.__support.HashTable.group_test.__unit__.__NO_DISABLE_SSE2_OPT
```
And check if the generated codes in `build/projects/libc/test/src/__support/HashTable/` match your expectation.

Thanks,

SchrodingerZhu replied:
I see. I forgot to remove that part.

lntue replied:
Do you mind adding a `TODO` at the test for me to investigating this?  I'm in the middle of re-implementing how the flags are controlled, and this will be a good use-case / example for that.

I think the PR is good to go, modulo this issue that I will investigate after merging.

Thanks,

lntue commented on libc/config/linux/x86_64/entrypoints.txt:
```diff
@@ -497,6 +497,14 @@ if(LLVM_LIBC_FULL_BUILD)
     libc.src.spawn.posix_spawn_file_actions_destroy
     libc.src.spawn.posix_spawn_file_actions_init
 
+    # search.h entrypoints
```
The only thing that requires `FULL_BUILD` is the generated headers that these entrypoints depend on.  But if they match with the system headers, then it's probably fine with the overlay mode.  You could try to clean, rebuild, and test these entrypoints in overlay mode to see if `FULL_BUILD` is needed.

nickdesaulniers commented on libc/src/__support/HashTable/table.h:
```diff
@@ -0,0 +1,235 @@
+    // for a successful search.
+    if (slot->key != nullptr) {
+      return slot;
+    }
```
please mark these threads as resolved once addressed.

gchatelet commented:
Thx @SchrodingerZhu for the patch.

As a side note I would really appreciate that such a big patch gets broken down into several pieces next time. Most importantly I think It's crucial that we are very careful about what we put in the `__support` folder. If we're not principled about this, we will soon be increasing our technical debt. And TBH I think it's already too much of a kitchen sink.

gchatelet commented:
Hi @SchrodingerZhu ,

Thx for your reply, I truly appreciate it.
Yes please don't hesitate to add me as a reviewer for PR touching files under `__support` or drop me an email if you want to discuss the approach beforehand. Thank you!
